### ISSUE TITLE
Design notifications for recurring team prompts

### ISSUE THREAD

**Issue by ackernaut** [ParabolInc/parabol#6075]
In early concepts we’ve painted in broad strokes how we may notify folks via chat or in the app, reminding them to participate in a stand-up, or notifying when somebody leaves a comments etc.

In this issue we want to refine what notifications we should and shouldn’t build, the finer details on how they work, and backlog the as enhancements.

### Open questions
- Do we decouple reminder notifications from literal meeting open and close times? Probably so. E.g. the daily stand-up is open until midnight. We don’t alert the team at 12:01 am. We let them pick times for reminder notifications. Maybe 1 or more?
- Do we offer specific types of notifications? Do we allow a custom message per reminder notification?
- Do we allow for sending participation reminders both as an in-app notif and a chat notif? One or the other?
- Do we alert the whole team when somebody initially responds to an activity? E.g. Terry added their async standup response
- Do we alert the team or the individuals involved when somebody comments on a thread? Or not at all? E.g. Aviva commented on your async standup response
- What is the first notification or so that we build for one-time team prompt activities (aka Standup MVP) or recurring team prompt activities?

### Acceptance criteria
- Designed concepts for what kind of notifications to support. What’s the meta and copy of each notification? What triggers the notification? How is it configured? Where might it land? In-app and/or chat to individual members? In-app and/or chat to team? 
- Proposed written criteria to be used for implementation
- Reviewed with a designer and developer

### Estimated effort
- 8 hours

**Comment by ackernaut** [comment]
cc @enriquesanchez added another issue! Questions? Anything missing?

**Comment by enriquesanchez** [comment]
> Do we offer specific types of notifications? 

What types of notifications do you have in mind?

Another thing that would be interesting to explore is grouping these notifications so users don't get a long stream of repetitive notifications:¡

- Aviva commented on your stand-up response
- Terry commented on your stand-up response
- Bruce reacted to your stand-up response
...and so on

Instead we could condense the noise into something a little more useful:

- Aviva, Terry and Bruce commented ans reacted on your stand-up response

**Comment by ackernaut** [comment]
@enriquesanchez 

Maybe we scope some time in this task for doing a quick audit of the existing notifications, say what should probably stay or not, and what might be added. The goal is not to design those, but really develop an opinion of what are the criteria of a good notification in our system, versus something that should be in chat, or on the timeline, etc.

Yes, your idea on rolling notifications into 1 is 💯 🙌

**Comment by ackernaut** [comment]
@enriquesanchez would you want to work on this one under the ‘notifications’ bucket?

**Comment by enriquesanchez** [comment]
Closed by #6423

**Comment by enriquesanchez** [comment]
## When a recurring activity starts

When a recurring activity starts (I'm using standups for this example), team members get a new notification. If the new [Activity view](https://github.com/ParabolInc/parabol/issues/6399) is live, it would look like this:

![image](https://user-images.githubusercontent.com/3276087/168132789-ac532ecf-8684-43bb-8a93-5416770abc41.png)

Otherwise, the notification will appear in the current Notifications menu. 

This notification confirms that the activity i**s now open**, and informs the deadline for the activity. It includes:

- icon for the activity
- name of recurring activity
- date of recurring activity (which is what distinguishes different instances of the same recurring activity)
- team to which the activity belongs to
- end time or deadline for the activity
- CTA to submit reponse 

A chat notification can also be generated, if the facilitator chose to notify via their chat integration, and it would look like this:

<img width="515" alt="start" src="https://user-images.githubusercontent.com/3276087/168129460-267bea16-236e-487a-9e49-65da2a96c052.png">


We use the same style and format we currently have for chat integration notifications. What's different is the content.

## Reminders

If reminders have been set for this recurring activity, both an activity view event and a chat notification get generated:

![image](https://user-images.githubusercontent.com/3276087/168132806-4e670ee4-80d0-45cf-91c4-abe710a7b64b.png)

<img width="758" alt="reminder" src="https://user-images.githubusercontent.com/3276087/168134367-f585dfbb-7adf-4e4e-a2d2-66a427c0e0b2.png">

The text informs that they still have time to submit a reponse for the current instance of the recurring activity, in this case `Daily Standup (May 11th)`

## Activity on your response

When there is new comment on your response, a notification gets generated:

![image](https://user-images.githubusercontent.com/3276087/168134788-3e510426-36e9-4853-b852-e2f105861149.png)

Besides the activity meta, the notification includes the author, timestamp and message, along with a CTA to view the comment.

If more comments come in, these notifications get grouped and are bubbled up in the timeline, with the most recent comments sitting on top:

![image](https://user-images.githubusercontent.com/3276087/168135114-9bd1c1a1-5eec-45c2-8cc5-abfb470e661a.png)

We can continue to group them and when there's more than 3, we collapse them and show a `...and X more` text button to expand and see all.

![image](https://user-images.githubusercontent.com/3276087/168135324-0af609a5-01d8-47f2-ba50-4d7502d9e037.png)

**Because these notifications are individual and not team wide, and because we don't yet have a way for users to set their notification preferences, I propose that for now these notifications will not be sent to any chat integration.** 

We can explore how individual user notification preferences could look like in a separate issue.

## Responses you follow

If you comment on someone else's response, you are automatically subscribed to that thread and in consequence get notified when new comments come in:

![image](https://user-images.githubusercontent.com/3276087/168135563-e2a408ec-7db6-425e-827c-136eb4984b6c.png)

Same as above, **because these notifications are individual and not team wide, and because we don't yet have a way for users to set their notification preferences, I propose that for now these notifications will not be sent to any chat integration.** 

We can explore in a separate issue for a way to subscribe and unsubscribe to someone else's response activity.

## When the instance ends

When the current instance of the recurring activity comes to and end, both a timeline/activity notification and a chat notification get generated:

![image](https://user-images.githubusercontent.com/3276087/168136365-48904568-ee4a-47dd-aaf5-a89d30ebb11e.png)

<img width="772" alt="end" src="https://user-images.githubusercontent.com/3276087/168137029-d323f981-c039-4a0a-80a9-555a12f17445.png">

The notification indicates that the instance is now closed, and provides a short summary of the activity along a CTA to view a summary.

# Next steps

- [ ] Have these concepts reviewed by a designer and developer
- [ ] Slice into smaller parts and write implementation issues
- [ ] Create any needed issues for future or out of scope design explorations

**Comment by enriquesanchez** [comment]
@BartoszJarocki @ackernaut Could I get a review and feedback from you on the proposal above?

**Comment by ackernaut** [comment]
@enriquesanchez taking a card!

**Comment by enriquesanchez** [comment]
@ackernaut @BartoszJarocki Pinging again for that review 🙈

**Comment by ackernaut** [comment]
Thanks for the reminder @enriquesanchez 

- I like that you took a broader pass on how all of these may tie together, so that we can break them down. Sometimes when smaller issues are written, more questions come up about the details, and can require smaller design efforts, too e.g. let’s suss out the details of bundling multiple comments. So these design efforts around broader patterns are good to time-box when the idea is high enough fidelity
- Today we have it set up so that an individual receives a private notification for ‘Meeting Time Limit Ended’ when they had set a time limit on a phase. We could probably explore a similar thing where they can add Slack, and check to receive private messages when folks comment on threads that involve them 1) it was my topic 2) I commented on a thread — in the future I might expect an obvious enhancement of being able to mute single threads
- I wonder how these thread comment notification designs would work for Sprint Poker? Probably very similar: 1) activity type 2) story being estimated = thread topic 3) the new comment on the thread. I think that I would see new comments when I had participated in a thread since I probably don’t ‘own’ the story like I do my reponse in a stand-up. The most likely cases are 1) I asked a question about a story or 2) I gave more information and somebody had a comment or question
- I wonder how we technically throttle bundles of comments? If I understood correctly, you’re suggesting the thread comment notification would be updated and brought to the top of the timeline? Or would we just give folks a latest comments every hour etc? I think the bundling and resurfacing pattern sounds smarter, more dynamic. In Slack I think the API allows you to do similar things, but it may take some exploration: can I update a previous private Slack notif, do they get pinged when it’s updated? can I throttle and bundle e.g. once per hour? do we have to push a single private message every time somebody comments?
- Overall I like the clean pattern you’re developing. It’s got just the right meta, the right content, and a single CTA 🙌

**Comment by ackernaut** [comment]
Edit: when you break this down into smaller issues, it would be good to capture a placeholder issue for ‘thread comment notifications for Sprint Poker’

**Comment by mattkrick** [comment]
@enriquesanchez can this issue be closed now? what is left to consider this marked as done?

**Comment by enriquesanchez** [comment]
@mattkrick Waiting on a review from @BartoszJarocki before this can be closed and implementation issues created. It's on his radar and there's no immediate rush 👍

**Comment by BartoszJarocki** [comment]
@enriquesanchez looks good! We have to keep in mind a couple things for now
1. We don't have deadlines yet, so we won't be showing that in notifications
2. We'll have to think about the mechanism of unsubscribing from selected thread notifications, similarly to slack (you commented on a thread, there's a big discussion going on but you're not interested in that anymore) 

I think this is more than enough to start drafting implementation issues for basic stuff like
1. Improving notification for starting the meeting
2. Adding notification when the meeting has ended
3. Thread notifications (for now probably just inside Parabol) 

Feel free to close this, we'll start creating implementation issues. good work!
